mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -2077,14 +2196,55 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
   populateSYCLToLLVMConversionPatterns(converter, patterns);
   populateFuncToLLVMConversionPatterns(converter, patterns);
 
-  ConversionTarget target(*context);
-  target.addIllegalDialect<sycl::SYCLDialect>();
-  // TODO: connect wires to lower this operation.
-  target.addLegalOp<sycl::SYCLLocalIDOp>();
-  target.addLegalDialect<LLVM::LLVMDialect>();
-  target.addLegalDialect<arith::ArithDialect>();
+  populateVectorToLLVMConversionPatterns(converter, patterns);
+  arith::populateArithToLLVMConversionPatterns(converter, patterns);
+  populateSPIRVToLLVMTypeConversion(converter);
+  populateSPIRVToLLVMConversionPatterns(converter, patterns);
+  populateSPIRVToLLVMFunctionConversionPatterns(converter, patterns);
+
+  LLVMConversionTarget target(context);
+  target.addDynamicallyLegalDialect<sycl::SYCLDialect>(
+      [](Operation *op) { return isa<sycl::SYCLLocalIDOp>(op); });
+
+  const auto res = applyPartialConversion(module, target, std::move(patterns));
+
+  LLVM_DEBUG(llvm::dbgs() << "Module after LLVM lowering:\n"; module.dump(););
+
+  return res;
+}
+
+void ConvertSYCLToLLVMPass::cleanUnrealizedConversionCasts() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "Reconciling UnrealizedConversionCasts operations...\n");
+
+  auto &context = getContext();
+  auto module = getOperation();
+
+  ConversionTarget target(context);
+  RewritePatternSet patterns(&context);
+
+  populateReconcileUnrealizedCastsPatterns(patterns);
+
+  if (applyPartialConversion(module, target, std::move(patterns)).failed()) {
+    LLVM_DEBUG(llvm::dbgs() << "WARNING: Could not remove every "
+                               "UnrealizedConversionCast operation\n");
+  }
+
+  LLVM_DEBUG(
+      llvm::dbgs()
+          << "Module after reconciling UnrealizedConversionCasts operations:\n";
+      module.dump(););
+}
 
-  target.addLegalOp<ModuleOp>();
-  if (failed(applyPartialConversion(module, target, std::move(patterns))))
-    signalPassFailure();
+void ConvertSYCLToLLVMPass::runOnOperation() {
+  constexpr int requiredRuns{2};
+  for (int i = 0; i < requiredRuns; ++i) {
```

Comment:
How do we know we need exactly 2 passes ? Also shouldn't we convert from SYCL to SPIRV first and then convert from to LLVM after ?

Reply:
Comment in file:
```
 2065 /// A pass converting MLIR SYCL operations into LLVM dialect.
 2066 ///
 2067 /// This pass relies on SYCL to GPU and target dialects, e.g., SPIRV, conversion
 2068 /// patterns. This pass is executed in 5 steps:
 2069 /// 1. Convert grid ops to a target dialect, e.g., SPIRV;
 2070 /// 2. Lower all of the operations to LLVM. As some of the operations will yield
 2071 ///    SYCL grid ops, we need to perform the same conversions as in step 1
 2072 ///    again;
 2073 /// 3. Same as 1;
 2074 /// 4. Lower remaining operations to LLVM;
 2075 /// 5. Resolve UnrealizedConversionCastOps appearing due to the fact that this
 2076 ///    pass is performed in several steps.
```